src/Pact/Types/KeySet.hs

```diff
@@ -114,29 +128,97 @@ instance Arbitrary KeySet where
 -- | allow `{ "keys": [...], "pred": "..." }`, `{ "keys": [...] }`, and just `[...]`,
 -- | the latter cases defaulting to "keys-all"
 instance FromJSON KeySet where
-    parseJSON v = withObject "KeySet" (\o ->
-                    KeySet <$> o .: "keys" <*>
-                    (fromMaybe defPred <$> o .:? "pred")) v <|>
-                  (KeySet <$> parseJSON v <*> pure defPred)
-      where defPred = Name (BareName "keys-all" def)
-instance ToJSON KeySet where
-    toJSON (KeySet k f) = object ["keys" .= k, "pred" .= f]
+    parseJSON v = withObject "KeySet" keyListPred v <|> keyListOnly
+      where
+        defPred = Name (BareName "keys-all" def)
+
+        keyListPred o = KeySet
+          <$> o .: "keys"
+          <*> (fromMaybe defPred <$> o .:? "pred")
 
+        keyListOnly = KeySet
+          <$> parseJSON v
+          <*> pure defPred
+
+instance ToJSON KeySet where
+    toJSON (KeySet k f) = object
+      [ "keys" .= k
+      , "pred" .= f
+      ]
 
 -- -------------------------------------------------------------------------- --
 -- KeySetName
 
-newtype KeySetName = KeySetName Text
-    deriving (Eq,Ord,IsString,AsString,ToJSON,FromJSON,Show,NFData,Generic,SizeOf)
+data KeySetName
+  = KeySetName
+  { _ksnName :: Text
+  , _ksnNamespace :: Maybe NamespaceName
+  } deriving (Eq, Ord, Show, Generic)
 
-instance Arbitrary KeySetName where
-  arbitrary = KeySetName <$> genBareText
-instance Pretty KeySetName where pretty (KeySetName s) = "'" <> pretty s
+instance IsString KeySetName where
+  fromString ksn = case parseAnyKeysetName (T.pack ksn) of
+    Left e -> error e
+    Right ks -> ks
 
+instance NFData KeySetName
+
+instance SizeOf KeySetName where
+  sizeOf (KeySetName ks nsn) =
+    constructorCost 2 + sizeOf ks + sizeOf nsn
+
+instance FromJSON KeySetName where
```

Comment:
the next two decoders need the whole forking trick with list `++`

Reply:
so this was derived before. Let's handle this in the sync